src/libespeak-ng/translate.c

```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+			idx++;
+		//idx++;
+	}
+	*from = idx;
+	if (**from == 0)
+		return false;
+
+	while (*idx != 0 && *idx != ' ') {   // get next 'to' position
+		idx++;
+	}
+	while (*idx == ' ') // skip remaining spaces, if exists
+		idx++;
+	//*to = ++idx;
+	*to = idx;
+	if (**to == 0)
+		return false;
+	return true;
+}
 
-	if (ignore_next) {
-		ignore_next = false;
-		return 8;
+static void ReplaceMatchingChars(char **matchstart, int *from, int *to, int upper_case)
+{
+	// Replace characters from 'from' to 'to' group in source clause, starting
+	// with place indicated in matchstart.
+	// Rewrites remaining part of source, if length of 'from' and 'to' groups differs.
+
+	char tmpsource[N_TR_SOURCE + 40]; // Prepare working copy of source
+	memcpy(&tmpsource, *matchstart, N_TR_SOURCE + 40);
+
+	// prepare from part
+	char charfrom[N_WORD_BYTES + 1] = { };
+	int usedfrom = 0;
+	int *pintfrom = from;
+	char *pcharfrom = &charfrom;
+	while (*pintfrom != ' ') {
+		usedfrom += utf8_out(*pintfrom, pcharfrom + usedfrom);
+		pintfrom++;
+	}
+
+	// prepare to part
+	char charto[N_WORD_BYTES + 1] = { };
+	int usedto = 0;
+	int *pintto = to;
+	char *pcharto = &charto;
+	while (*pintto != 0 && *pintto != ' ') {
+		usedto += utf8_out(*pintto, pcharto + usedto);
+		pintto++;
+	}
+
+	// Set up replacement in temporary buffer
+	char *psource = *matchstart;
+	char *ptmpsource = &tmpsource;
+	pcharto = &charto;
+	int replace = 0;
+
+	// Do replacement writing 'to' part in buffer
+	while (*pcharto != 0 ) {
+		*ptmpsource = *pcharto;
+		pcharto++;
+		ptmpsource++;
+	}
+
+	// If 'to' part is different than 'from' part, fix remaining part of the buffer.
+	// (Read after 'from' in source, write after 'to' in buffer.
+	// Don't touch last 40 bytes of source, which could contain control data.
+	if (usedfrom != usedto) {
+		psource = *matchstart + usedfrom;
+		while (*psource != 0 && psource < &source[N_TR_SOURCE]
+				&& ptmpsource < &tmpsource[N_TR_SOURCE]) {
+			*ptmpsource = *psource;
+			ptmpsource++;
+			psource++;
+		}
 	}
-	if (c == 0) return 0;
-
-	if ((replace_chars = tr->langopts.replace_chars) == NULL)
-		return c;
 
-	// there is a list of character codes to be substituted with alternative codes
-
-	if (iswupper(c_lower = c)) {
-		c_lower = towlower2(c, tr);
-		upper_case = 1;
+	// Write temporary buffer back to source
+	psource = *matchstart;
+	ptmpsource = &tmpsource;
+	while (*ptmpsource != 0 && ptmpsource < &tmpsource[N_TR_SOURCE]
+			&& psource < &source[N_TR_SOURCE]) {
+		*psource = *ptmpsource;
+		ptmpsource++;
+		psource++;
 	}
 
-	new_c = 0;
-	for (ix = 0; (word = replace_chars[ix]) != 0; ix += 2) {
-		if (c_lower == (word & 0xffff)) {
-			if ((word >> 16) == 0) {
-				new_c = replace_chars[ix+1];
-				break;
-			}
-			if ((word >> 16) == (unsigned int)towlower2(next_in, tr)) {
-				new_c = replace_chars[ix+1];
-				ignore_next = true;
-				break;
-			}
-		}
+	// if requested, print trace
+	if (option_phonemes & espeakPHONEMES_TRACE) {
```

Comment:
[Minor] A single-statement if statement does not need braces. This is to keep a consistent coding style with the rest of the codebase.